utils/zerovec/src/hashmap/algorithms.rs

```diff
@@ -0,0 +1,167 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use alloc::borrow::Borrow;
+use alloc::vec;
+use alloc::vec::Vec;
+use core::hash::{Hash, Hasher};
+use t1ha::T1haHasher;
+
+// Const seed to be used with [`T1haHasher::with_seed`].
+const SEED: u64 = 0xaabbccdd;
+
+/// Split the 64bit `hash` into (g, f0, f1).
+/// g denotes the highest 16bits of the hash modulo `m`, and is referred to as first level hash.
+/// (f0, f1) denotes the middle, and lower 24bits of the hash respectively.
+/// (f0, f1) are used to distribute the keys with same g, into distinct slots.
+///
+/// # Arguments
+///
+/// * `hash` - The hash to split.
+/// * `m` - The modulo used to split the hash.
+#[inline]
+pub const fn split_hash64(hash: u64, m: u32) -> (usize, u32, u32) {
+    (
+        ((hash >> 48) as u32 % m) as usize,
+        (hash >> 24) as u32 & 0xffffff,
+        ((hash & 0xffffff) as u32),
+    )
+}
+
+/// Compute hash using [`T1haHasher`].
+#[inline]
+pub fn compute_hash<K: Hash + ?Sized>(key: &K) -> u64 {
+    let mut hasher = T1haHasher::with_seed(SEED);
+    key.borrow().hash(&mut hasher);
+    hasher.finish()
+}
+
+/// Calculate the index using (f0, f1), (d0, d1) in modulo m.
+/// Returns [`None`] if d is (0, 0) or modulo is 0
+/// else returns the index computed using (f0 + f1 * d0 + d1) mod m.
+#[inline]
+pub fn compute_index(f: (u32, u32), d: (u32, u32), m: u32) -> Option<usize> {
+    if d == (0, 0) || m == 0 {
+        None
+    } else {
+        Some((f.1.wrapping_mul(d.0).wrapping_add(f.0).wrapping_add(d.1) % m) as usize)
```

Comment:
Reiterating my previous comment because the code moved: `wrapping_` does (mod 2^32) arithmetic, but you want (mod m).

Reply:
Done

Reply:
Reverted back to `wrapping_mul` and `wrapping_add`.